opm/simulators/wells/BlackoilWellModelGuideRates.cpp

```diff
@@ -0,0 +1,588 @@
+
+#include <opm/output/data/Groups.hpp>
+#include <opm/output/data/GuideRateValue.hpp>
+
+#include <opm/simulators/wells/BlackoilWellModelGeneric.hpp>
+#include <opm/simulators/wells/WellGroupHelpers.hpp>
+#include <opm/simulators/wells/WellInterfaceGeneric.hpp>
+
+#include <fmt/format.h>
+
+namespace {
+
+struct RetrieveWellGuideRate
+{
+    RetrieveWellGuideRate() = default;
+
+    explicit RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                   const std::string&    wgname);
+
+    explicit RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                   const Opm::Group&     group);
+
+    bool prod      { false };
+    bool inj_water { false };
+    bool inj_gas   { false };
+};
+
+RetrieveWellGuideRate
+operator||(RetrieveWellGuideRate lhs, const RetrieveWellGuideRate& rhs)
+{
+    lhs.prod      = lhs.prod      || rhs.prod;
+    lhs.inj_water = lhs.inj_water || rhs.inj_water;
+    lhs.inj_gas   = lhs.inj_gas   || rhs.inj_gas;
+
+    return lhs;
+}
+
+RetrieveWellGuideRate::RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                             const std::string&    wgname)
+    : prod      { guideRate.has(wgname) }
+    , inj_water { guideRate.has(wgname, Opm::Phase::WATER) }
+    , inj_gas   { guideRate.has(wgname, Opm::Phase::GAS)   }
+{}
+
+RetrieveWellGuideRate::RetrieveWellGuideRate(const Opm::GuideRate& guideRate,
+                                             const Opm::Group&     group)
+    : RetrieveWellGuideRate{ guideRate, group.name() }
+{
+    if (group.isProductionGroup()) {
+        this->prod = true;
+    }
+
+    if (group.isInjectionGroup()) {
+        this->inj_water = this->inj_water || group.hasInjectionControl(Opm::Phase::WATER);
+        this->inj_gas   = this->inj_gas   || group.hasInjectionControl(Opm::Phase::GAS);
+    }
+}
+
+class GroupTreeWalker
+{
+public:
+    using GroupOp = std::function<void(const Opm::Group&)>;
+    using WellOp = std::function<void(const Opm::Well&)>;
+
+    explicit GroupTreeWalker(const Opm::Schedule& sched,
+                             const int            reportStepIdx)
+        : sched_        (sched)
+        , reportStepIdx_(reportStepIdx)
+    {}
+
+    GroupTreeWalker& groupOp(GroupOp visit)
+    {
+        this->visitGroup_ = std::move(visit);
+        return *this;
+    }
+
+    GroupTreeWalker& wellOp(WellOp visit)
+    {
+        this->visitWell_ = std::move(visit);
+        return *this;
+    }
+
+    void clear()
+    {
+        this->visitGroup_ = GroupOp{};
+        this->visitWell_ = WellOp{};
+    }
+
+    void traversePreOrder();
+    void traversePostOrder();
+
+private:
+    using NodeOp = void (GroupTreeWalker::*)(std::string_view) const;
+
+    std::reference_wrapper<const Opm::Schedule> sched_;
+    int reportStepIdx_;
+
+    GroupOp visitGroup_{};
+    WellOp visitWell_{};
+
+    std::stack<std::string_view, std::vector<std::string_view>> dfsGroupStack_{};
```

Comment:
We must include `<stack>` in order to use this type.

Reply:
done